convert/gitlab/yaml/job.go

```diff
@@ -19,75 +19,73 @@ import "errors"
+	Retry             *Retry                   `yaml:"retry,omitempty"`
+	Rules             []*Rule                  `yaml:"rules,omitempty"`
+	Script            Stringorslice            `yaml:"script,omitempty"`
+	Secrets           map[string]*Secret       `yaml:"secrets,omitempty"`
+	Services          []*Image                 `yaml:"services,omitempty"`
+	Stage             string                   `yaml:"stage,omitempty"`
+	Tags              Stringorslice            `yaml:"tags,omitempty"`
+	Timeout           string                   `yaml:"timeout,omitempty"`
+	Trigger           *Trigger                 `yaml:"trigger,omitempty"`
+	Variables         map[string]*Variable     `yaml:"variables,omitempty"`
+	When              string                   `yaml:"when,omitempty"` // on_success, manual, always, on_failure, delayed, never
 }
 
-// UnmarshalYAML implements the unmarshal interface.
-func (v *Job) UnmarshalYAML(unmarshal func(interface{}) error) error {
-	var out1 []string
-	var out2 = struct {
-		After         Stringorslice        `yaml:"after_script,omitempty"`
-		Artifacts     *Artifacts           `yaml:"artifacts,omitempty"`
-		AllowFailure  *AllowFailure        `yaml:"allow_failure,omitempty"`
-		Before        Stringorslice        `yaml:"before_script,omitempty"`
-		Cache         *Cache               `yaml:"cache,omitempty"`
-		Environment   *Environment         `yaml:"environment,omitempty"`
-		Extends       Stringorslice        `yaml:"extends,omitempty"`
-		Image         *Image               `yaml:"image,omitempty"`
-		Inherit       *Inherit             `yaml:"inherit,omitempty"`
-		Interruptible bool                 `yaml:"interruptible,omitempty"`
-		Needs         []*Needs             `yaml:"needs,omitempty"`
-		Pages         *Job                 `yaml:"pages,omitempty"`
-		Parallel      *Parallel            `yaml:"parallel,omitempty"`
-		Release       *Release             `yaml:"release,omitempty"`
-		ResourceGroup string               `yaml:"resource_group,omitempty"`
-		Retry         *Retry               `yaml:"retry,omitempty"`
-		Rules         []*Rule              `yaml:"rules,omitempty"`
-		Script        Stringorslice        `yaml:"script,omitempty"`
-		Secrets       map[string]*Secret   `yaml:"secrets,omitempty"`
-		Services      []*Image             `yaml:"services,omitempty"`
-		Stage         string               `yaml:"stage,omitempty"`
-		Tags          Stringorslice        `yaml:"tags,omitempty"`
-		Timeout       string               `yaml:"timeout,omitempty"`
-		Trigger       *Trigger             `yaml:"trigger,omitempty"`
-		Variables     map[string]*Variable `yaml:"variables,omitempty"`
-		When          string               `yaml:"when,omitempty"`
-	}{}
+type DASTConfiguration struct {
+	SiteProfile    string `yaml:"site_profile,omitempty"`
+	ScannerProfile string `yaml:"scanner_profile,omitempty"`
+}
+
+type IDToken struct {
+	Aud interface{} `yaml:"aud"`
+}
+
+type NeedsListOrStruct struct {
```

Comment:
move to a `needs.go` file

we can simplify the struct to this:

```
type Needs struct {
  Items []*Need
}


type Need struct {
   ...
}
```

see this example travis code for modeling this pattern:
https://github.com/drone/go-convert/blob/8e7933be134543860b2e48adad2ae63c9c40190c/convert/travis/yaml/addons_apt.go#L74C1-L74C1

Reply:
also note that this will probably need a custom yaml marshal function to ensure it produces a proper GitLab yaml when marshaled

Reply:
done.